### PR TITLE
Go To Button

### PR DESCRIPTION
**Pull request by luc-ass**
### Added
- configuration for optional button with: name, x-coordinate, y-coordinate
- logic to send robot vacuum to location
- explanation to make finding the right coordinates somewhat easier
### Known quirks
- You need to start/stop the vacuum to send it back to its dock. My way is actually pressing the dock button on the vacuum, as I send it to the bin anyway...
- The default name of the button is highly opinionated - after all that's why I built this 😅 

Please let me know if I can improve it somehow.

### CODE REVIEW

**Comment by luc-ass** [comment]
This resolves #306, #256 and #130.

File: index.js [hunk 1]
```diff
@@ -14,6 +14,7 @@ const PLUGIN_NAME = "homebridge-xiaomi-roborock-vacuum";
 const ACCESSORY_NAME = "XiaomiRoborockVacuum";
 
 const MODELS = require("./models");
+const { runInThisContext } = require("vm");
```
Comment (luc-ass) [inline index.js:17]: no idea where this came from

File: miio/lib/devices/vacuum.js [hunk 2]
```diff
@@ -332,6 +332,14 @@ module.exports = class extends Vacuum.with(
     return this.call("find_me", [""]).then(() => null);
   }
 
+  /**
+   * Send the vacuum to a specific location.
+  */
+  sendToLocation(x,y) {
+    console.log("called sendToLocation()")
```
Comment (luc-ass) [inline miio/lib/devices/vacuum.js:339]: might be removed if not needed

**Comment by luc-ass** [comment]
Would this be a solution for the "button always off" problem?

https://github.com/homebridge-xiaomi-roborock-vacuum/homebridge-xiaomi-roborock-vacuum/blob/master/miio/lib/devices/vacuum.js#L225-L230

File: index.js (hunk 1 above)
Comment (luc-ass) [reply index.js]: resolved

File: miio/lib/devices/vacuum.js (hunk 2 above)
Comment (luc-ass) [reply miio/lib/devices/vacuum.js]: ✅ resolved

File: index.js [hunk 3]
```diff
@@ -1604,6 +1604,24 @@ class XiaomiRoborockVacuum {
...
+      return goingToLocation;
+    } catch (err) {
+      this.log.error(
+        `ERR getGoToState | ${this.model} | Failed getting the cleaning status.`,
+        err
+      );
+      throw err;
+    }
+  }
+
```
Comment (luc-ass) [inline index.js:1624]: does now show the button as `on` while the vacuum is driving towards location

File: miio/lib/devices/vacuum.js [hunk 4]
```diff
@@ -77,6 +77,8 @@ module.exports = class extends Vacuum.with(
           return "updating";
         case 15:
           return "docking";
+        case 16:
+          return "going-to-location"
```
Comment (luc-ass) [inline miio/lib/devices/vacuum.js:81]: 16: pin&go

File: index.js [hunk 5]
```diff
@@ -1576,6 +1591,37 @@ class XiaomiRoborockVacuum {
...
+    try {
+      await this.device.sendToLocation(this.config.goToX,this.config.goToY);
+      callback();
+    } catch (err) {
+      this.log.error(`ERR goTo | ${this.model} | `, err);
+      callback(err);
+    }
+  }
+
+  async getGoToState() {
```
Comment (luc-ass) [inline index.js:1607]: Added feedback for the button. `on` while moving to target.

File: config.schema.json [hunk 6]
```diff
@@ -69,6 +69,39 @@
...
+        "default": false,
+        "required": true
+      },
+      "goToWord": {
+        "title": "Go To Word",
+        "type": "string",
+        "default": "time to empty",
+        "required": true,
+        "condition": {
+          "functionBody": "return model.goTo === true"
```
Comment (luc-ass) [inline config.schema.json:84]: only show advanced configuration if `go to` is enabled

Comment (luc-ass) [reply config.schema.json:84]: I will happily add this to other parts of the configuration, if you wish 🎁

File: README.md [hunk 7]
```diff
@@ -154,6 +154,13 @@ This feature seems to be working on models that support naming the rooms in the
 
 Even if you have one of these models but you haven't named the Rooms in your App yet, this function will not work! Thanks @domeOo
 
+## Go to Location
+This plugin allowes you to send your vacuum to a specific point on the map. As there is no easy way of figuring out the coordinates of your desired location, here are some steps to help you find the right spot.
```
Comment (afharo) [inline README.md:158]: typo 😇 
```suggestion
This plugin allows you to send your vacuum to a specific point on the map. As there is no easy way of figuring out the coordinates of your desired location, here are some steps to help you find the right spot.
```

File: README.md [hunk 8]
```diff
@@ -154,6 +154,13 @@ This feature seems to be working on models that support naming the rooms in the
...
+1. The base location (typically the position of the dock when the map was created) is at `[25500,25500]` where these values are in the format `[x-axis/horizontal, y-axis/vertical]`.
+2. Increasing the first value results in going to the right, increasing the second value results in going to the top of the map (as displayed inside the app). The value seems to be in Millimeters.
```
Comment (afharo) [inline README.md:160]: Reading points 1 and 3 makes me wonder if it should be
```suggestion
2. Increasing the first value results in going to the top of the map, increasing the second value results in going to the right of the map (as displayed inside the app). The value seems to be in Millimeters.
```

Did I miss something?

File: index.js [hunk 9]
```diff
@@ -106,6 +106,10 @@ class XiaomiRoborockVacuum {
     this.config.pauseWord = config.pauseWord || "Pause";
     this.config.findMe = config.findMe || false;
     this.config.findMeWord = config.findMeWord || "where are you";
+    this.config.goTo = config.goTo || false;
+    this.config.goToWord = config.goToWord || "time to empty";
```
Comment (afharo) [inline index.js:110]: I love your use case, but probably `Go to coordinates` is more generic to everyone :)

**Review by afharo — COMMENTED** [review]
Wow! This is a very requested feature! Thank you @luc-ass, for taking the time to implement it 🧡 

I added a couple of minor comments.

I also have a small suggestion if you think you can make it:
Is there a possibility to create multiple `Go To` buttons? Do you think it's worth implementing this feature that way?

File: index.js (hunk 9 above)
Comment (luc-ass) [reply index.js]: Good idea. I was out of ideas yesterday evening 😉

**Comment by luc-ass** [comment]
> Wow! This is a very requested feature! Thank you @luc-ass, for taking the time to implement it 🧡

Sure thing! Can't wait to use the automation myself!
> I added a couple of minor comments.

Thank you for your suggestions. I implemented them.
> I also have a small suggestion if you think you can make it: Is there a possibility to create multiple `Go To` buttons? Do you think it's worth implementing this feature that way?

As of now I have no idea how to do that in an efficient way. But I will happily keep working on it when I find some more time. 😉

**Comment by afharo** [comment]
We can merge this for now with one button and work on an improvement in the future. How does that sound?

**Review by afharo — APPROVED** [review]
LGTM! Thank you @luc-ass!

I'll go ahead, merge it and publish it.

We can have a follow up in the future to add multiple buttons.

**Pull request merged**